academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1559 @@
+        """
+
+        updatefiles_datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        success = download_datafiles(
+            datafile_list=updatefiles_datafiles,
+            ftp_server_url=self.ftp_server_url,
+            ftp_port=self.ftp_port,
+            reset_ftp_counter=self.reset_ftp_counter,
+            max_download_retry=self.max_download_retry,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def upload_downloaded_updatefiles(self, release: PubMedRelease, **kwargs):
+        """Upload downloaded updatefiles files to GCS."""
+
+        datafiles_to_upload = [
+            datafile.download_file_path for datafile in release.datafile_list if not datafile.baseline
+        ]
+
+        success = gcs_upload_files(
+            bucket_name=self.cloud_workspace.download_bucket,
+            file_paths=datafiles_to_upload,
+        )
+
+        set_task_state(success, kwargs["ti"].task_id, release=release)
+
+    def transform_updatefiles(self, release: PubMedRelease, **kwargs):
+        """
+        Transform the *.xml.gz files downloaded from PubMed's FTP server into usable json-like files for BigQuery import.
+
+        This is a multithreaded and pulls the PubmedArticle records from the downloaded XML files.
+        """
+
+        # Get list of datafiles for the baseline to trasform.
+        datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
+
+        # Object to store all of the upserts and delete keys that are present in each file.
+        record_keys = {}
+
+        # Process files in batches so that ProcessPoolExecutor doesn't deplete the system of memory
+        for i, chunk in enumerate(get_chunks(input_list=datafiles, chunk_size=self.max_processes)):
+            with ProcessPoolExecutor(max_workers=self.max_processes) as executor:
+                logging.info(f"In chunk {i} and processing files: {chunk}")
+
+                futures = []
+                datafile: Datafile
+                for datafile in chunk:
+                    input_path = datafile.download_file_path
+                    upsert_path = datafile.transform_upsert_file_path
+                    delete_path = datafile.transform_delete_file_path
+
+                    futures.append(executor.submit(transform_pubmed, input_path, upsert_path, delete_path))
+
+                for future in as_completed(futures):
+                    filename, record_keys_per_file = future.result()
+                    record_keys[filename] = record_keys_per_file
+
+        assert len(datafiles) == len(
+            record_keys.keys()
+        ), f"Number of updatefiles does not match the number of keys in record_keys: {len(datafiles)} vs {len(record_keys.keys())}"
+
+        ti: TaskInstance = kwargs["ti"]
+        ti.xcom_push(key="record_keys", value=record_keys)
+
+    def merge_upserts_and_deletes(self, release: PubMedRelease, **kwargs):
```

Comment:
Whilst the output of this function is correct, it could be much more succinct:

```python
import dataclasses

@dataclasses.dataclass
class PMID:
    value: int
    version: int

    @property
    def pmid(self):
        return f"{self.value}-{self.version}"

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return False
        return self.value == other.value and self.version == other.version

    def __hash__(self):
        return hash(self.pmid)


@dataclasses.dataclass
class PubmedUpdatefile:
    name: str
    upserts: List[PMID]
    deletes: List[PMID]

    @property
    def sequence(self):
        return int(re.search(r"(?<=pubmed)(\d+)(?=n)", self.name).group(1))

    @property
    def year(self):
        return int(re.search(r"(?<=n)(\d+)", self.name).group(1))

    def __eq__(self, other):
        if not isinstance(other, self.__class__):
            return False
        return self.name == other.name and set(self.upserts) == set(self.deletes)


def merge_upserts_and_deletes(updatefiles: List[PubmedUpdatefile]) -> Tuple[Dict[PMID, str], Set[PMID]]:
    upsert_index = dict()
    deletes = set()

    updatefiles.sort(key=lambda x: x.name)
    for file in updatefiles:
        # Apply upserts
        for upsert in file.upserts:
            # Check if item was previously deleted
            # and remove this delete if so
            if upsert in deletes:
                deletes.remove(upsert)

            # Update the file where the upsert comes from
            upsert_index[upsert] = file.name

        # Apply deletes
        for delete in file.deletes:
            # Delete any previous upserts
            if delete in upsert_index:
                del upsert_index[delete]

            # Add to delete set, as the record could still be in the main table
            deletes.add(delete)

    return upsert_index, deletes
```

The output of the function is the `upsert_index` dictionary, which you can directly use when iterating over the real updatefiles to determine which records to write out, only saving the upserts where the PMID key returns the same update file name.

You can also just write out the PMIDs in the deletes set straight after this function into one jsonl file, that doesn't need to be done in `save_pubmed_merged_records`, which can just be responsible for writing out the reconciled update files.

I would make the updatefile and PMID into a set of dataclasses and separate out the logic for merging the updates and deletes into its own function separate from the workflow so that it is easy to test and write a more in depth test case:
```python
def test_merge_upserts_and_deletes(self):
    changefiles = [
        PubmedUpdatefile(
            "pubmed10n2023",
            # Insert new records
            [PMID(10, 1), PMID(11, 1), PMID(12, 1), PMID(13, 1), PMID(14, 1), PMID(15, 1)],
            # Delete records from baseline
            [PMID(1, 1), PMID(2, 1)],
        ),
        PubmedUpdatefile(
            "pubmed11n2023",
            # Upsert over a previous record: PMID(10, 1)
            # Add a new version: PMID(12, 2), PMID(13, 2)
            # Add a new record: PMID(16, 1), PMID(17, 1)
            [PMID(10, 1), PMID(12, 2), PMID(13, 2), PMID(16, 1), PMID(17, 1)],
            # Delete a record from the previous day: PMID(11, 1)
            # Delete a record that was upserted on the same day: PMID(18, 1) n.b. this is removed from upserts
            [PMID(11, 1), PMID(18, 1)],
        ),
        PubmedUpdatefile(
            "pubmed12n2023",
            # Upsert over a previous record: PMID(10, 1), PMID(17, 1)
            # Add new records:
            [PMID(10, 1), PMID(17, 1), PMID(19, 1), PMID(20, 1), PMID(21, 1)],
            # Delete record added on previous day: PMID(13, 2)
            [PMID(13, 2)],
        ),
        PubmedUpdatefile(
            "pubmed13n2023",
            # Add a record deleted on a previous day: PMID(13, 2)
            # Add new versions:
            # Add new records:
            [PMID(13, 2), PMID(17, 1), PMID(19, 1), PMID(22, 1), PMID(23, 1)],
            # Delete records:
            [],
        ),
        PubmedUpdatefile(
            "pubmed14n2023",
            # Upsert new records:
            [PMID(24, 1), PMID(25, 1)],
            # Delete records that were previously upserted multiple times: PMID(17, 1)
            [PMID(17, 1)],
        ),
    ]
    expected_upserts = {
        PMID(12, 1): "pubmed10n2023",
        PMID(13, 1): "pubmed10n2023",
        PMID(14, 1): "pubmed10n2023",
        PMID(15, 1): "pubmed10n2023",
        PMID(12, 2): "pubmed11n2023",
        PMID(16, 1): "pubmed11n2023",
        PMID(10, 1): "pubmed12n2023",
        PMID(20, 1): "pubmed12n2023",
        PMID(21, 1): "pubmed12n2023",
        PMID(13, 2): "pubmed13n2023",
        PMID(19, 1): "pubmed13n2023",
        PMID(22, 1): "pubmed13n2023",
        PMID(23, 1): "pubmed13n2023",
        PMID(24, 1): "pubmed14n2023",
        PMID(25, 1): "pubmed14n2023",
    }
    expected_deletes = {PMID(1, 1), PMID(2, 1), PMID(11, 1), PMID(18, 1), PMID(17, 1)}

    # Check if we receive expected output
    actual_upserts, actual_deletes = merge_upserts_and_deletes(changefiles)
    self.assertEqual(expected_upserts, actual_upserts)
    self.assertEqual(expected_deletes, actual_deletes)
```

Then you can consolidate the tasks: `self.merge_upserts_and_deletes` and `self.save_merged_upserts_and_deletes` into one task.

Reply:
So the final `merge_upserts_and_deletes` task would look something like:

```python
def merge_upserts_and_deletes(self, release: PubMedRelease, **kwargs):
    ti: TaskInstance = kwargs["ti"]
    updatefiles: List[PubmedUpdatefile] = parse_updatefiles(ti.xcom_pull(key="updatefiles"))

    # Merge records
    upsert_index, deletes = merge_upserts_and_deletes(updatefiles)

    # Save deletes
    save_jsonl_gz("/path/to/deletes.jsonl.gz", [delete.to_dict() for delete in deletes])

    # Save final upserts
    datafiles = [datafile for datafile in release.datafile_list if not datafile.baseline]
    for i, chunk in enumerate(get_chunks(input_list=datafiles, chunk_size=self.max_processes)):
        with ProcessPoolExecutor(max_workers=self.max_processes) as executor:
            logging.info(f"In chunk {i} and processing files: {chunk}")

            futures = {}
            for datafile in chunk:
                filename = os.path.basename(datafile.download_file_path)
                assert (
                    filename in upsert_index
                ), f"records for file: {filename} are not present in upsert_index. "

                future = executor.submit(
                    save_pubmed_merged_records,
                    upsert_index,
                    datafile.transform_upsert_file_path,
                    datafile.merged_upsert_file_path,
                )
                futures[future] = filename

            for future in as_completed(futures):
                filename = futures[future]
                logging.info(f"Finished writing out upserts and deletes for updatefile: {filename}")
```